aws_advanced_python_wrapper/host_monitoring_v2_plugin.py

```diff
@@ -0,0 +1,506 @@
+                            monitoring_properties[property_key]
+                        monitoring_properties.pop(property_key, None)
+
+                logger.debug("HostMonitorV2.OpeningMonitoringConnection", self._host_info.url)
+                self._monitoring_connection = self._plugin_service.force_connect(self._host_info, monitoring_properties)
+                logger.debug("HostMonitorV2.OpenedMonitoringConnection", self._host_info.url)
+                return True
+
+            valid_timeout = ((self._failure_detection_interval_ns - self._THREAD_SLEEP_NANO) // 2) * 10**9
+            return self._is_host_available(self._monitoring_connection, valid_timeout)
+        except AwsWrapperError:
+            return False
+        finally:
+            if connect_telemetry_context is not None:
+                connect_telemetry_context.close_context()
+
+    def _is_host_available(self, conn: Connection, timeout_sec: float) -> bool:
+        try:
+            self._execute_conn_check(conn, timeout_sec)
+            return True
+        except TimeoutError:
+            return False
+        except psycopg.errors.ConnectionTimeout:
+            return False
+        except psycopg.OperationalError:
+            return False
+
+    def _execute_conn_check(self, conn: Connection, timeout_sec: float):
+        driver_dialect = self._plugin_service.driver_dialect
+        with conn.cursor() as cursor:
+            query = HostMonitorV2._QUERY
+            driver_dialect.execute("Cursor.execute", lambda: cursor.execute(query), query, exec_timeout=timeout_sec)
+            cursor.fetchone()
+
+    def _update_node_health_status(
+            self,
+            connection_valid: bool,
+            status_check_start_ns: float,
+            status_check_end_ns: float) -> None:
+        if not connection_valid:
+            self._failure_count += 1
+
+            if self._invalid_node_start_time_ns == 0:
+                self._invalid_node_start_time_ns = int(status_check_start_ns)
+
+            invalid_node_duration_ns = status_check_end_ns - self._invalid_node_start_time_ns
+            max_invalid_node_duration_ns = (
+                    self._failure_detection_interval_ns * max(0, self._failure_detection_count - 1))
+
+            if invalid_node_duration_ns >= max_invalid_node_duration_ns:
+                logger.debug("HostMonitorV2.HostDead", self._host_info.host)
+                self._is_unhealthy = True
+                return
+
+            logger.debug("HostMonitorV2.HostNotResponding", self._host_info.host, self._failure_count)
+
+        if self._failure_count > 0:
+            # Node is back alive
+            logger.debug("HostMonitorV2.HostAlive", self._host_info.host)
+
+        self._failure_count = 0
+        self._invalid_node_start_time_ns = 0
+        self._is_unhealthy = False
+
+    def abort_connection(self, connection: Connection) -> None:
+        try:
+            connection.close()
```

Comment:
1. Does this need to be wrapped in a timeout? I'm not sure if it would freeze indefinitely if the network was down
2. Should we call dialect.abort_connection here instead?